### PR TITLE
Add temp OID support for type, index, and constraint creation

### PR DESCRIPTION
**Pull request by time-change**
### Description

Normal OIDs were still being consumed in some cases during temp table creation even with temp oid buffer on. To fix this, ensure that the appropriate temp table dependent objects also use temp OID.

1. Redirect type creation functions to use temp OID when it's a TSQL temp table.
2. Redirect constraint creation to use temp OID in those cases as well. 
3. Remove workaround for indexes on temp tables. Previously, we needed to force indexes on temp tables to consume normal OIDs. With changes described in lines 1+2 above, we can now remove the workaround added in #232 and properly allow indexes to use temp oid buffer. 

Extension PR: babelfish-for-postgresql/babelfish_extensions#2443
 
### Issues Resolved

BABEL-4750
 
### Check List

- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/utils/misc/queryenvironment.c [hunk 1]
```diff
@@ -328,7 +329,8 @@ bool ENRgetSystableScan(Relation rel, Oid indexId, int nkeys, ScanKey key, List
 		reloid != DependRelationId &&
 		reloid != SharedDependRelationId &&
 		reloid != IndexRelationId &&
-		reloid != SequenceRelationId)
+		reloid != SequenceRelationId &&
+		reloid != AttrDefaultRelationId)
```
Comment (lejaokri) [inline src/backend/utils/misc/queryenvironment.c]: Adding `pg_attrdef` in ENR would look like we regress the following. Have we decided we are okay to take this hit?

```
[clouddesk3] % sqlcmd -S localhost -U jdbc_user -P 1 -e -i ~/test2.sql
CREATE TABLE #temptable(ID INT, ProductName varchar(20) DEFAULT 'product')
INSERT into #temptable (ID) VALUES (1)


(1 rows affected)

BEGIN TRANSACTION
ALTER TABLE #temptable ADD CONSTRAINT df_ProductName DEFAULT 'service' FOR ProductName
INSERT into #temptable (ID) VALUES (2)
select * from #temptable
ROLLBACK


(1 rows affected)
ID          ProductName
----------- --------------------
          1 product
          2 service

(2 rows affected)

INSERT into #temptable (ID) VALUES (3)
select * from #temptable


(1 rows affected)
ID          ProductName
----------- --------------------
          1 product
          3 service. --> should be product because of ROLLBACK

(2 rows affected)
```

Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: I think, since this is a byproduct of the ROLLBACK issue we have discussed, I would rather merge `pg_attrdef` now rather than block this entire change until the ROLLBACK issues are fixed. It would be better to space out large changes to ENR rather than merging large changes near the end of the release cycle

Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: Resolving as ENR Rollback is fixed now

**Comment by time-change** [comment]
Note: Hold off on merging until Temp OID blockers are fixed.

File: src/backend/catalog/heap.c [hunk 2]
```diff
@@ -1409,8 +1414,26 @@ heap_create_with_catalog(const char *relname,
...
+
+			/* We should not be creating TSQL Temp Tables in pg_upgrade. */
+			Assert(!IsBinaryUpgrade);
+
+			pg_type = table_open(TypeRelationId, AccessShareLock);
+			new_array_oid = GetNewTempOidWithIndex_hook(pg_type, TypeOidIndexId, Anum_pg_type_oid);
+			
+			Assert(!OidIsValid(reltypeid));
+
+			reltypeid = GetNewTempOidWithIndex_hook(pg_type, TypeOidIndexId, Anum_pg_type_oid);
```
Comment (lejaokri) [inline src/backend/catalog/heap.c]: we do not do this in the normal case. can you add a comment why?

Comment (time-change) [reply src/backend/catalog/heap.c]: done

File: src/backend/catalog/heap.c [hunk 3]
```diff
@@ -1428,6 +1456,7 @@ heap_create_with_catalog(const char *relname,
 										   ownerid,
 										   reltypeid,
 										   new_array_oid);
+
```
Comment (surendravishnoi) [inline src/backend/catalog/heap.c]: unnecessary change

File: src/backend/catalog/pg_constraint.c [hunk 4]
```diff
@@ -172,8 +177,17 @@ CreateConstraintEntry(const char *constraintName,
 		values[i] = (Datum) NULL;
 	}
 
-	conOid = GetNewOidWithIndex(conDesc, ConstraintOidIndexId,
-								Anum_pg_constraint_oid);
+	if (sql_dialect == SQL_DIALECT_TSQL && isTempNamespace(constraintNamespace))
```
Comment (surendravishnoi) [inline src/backend/catalog/pg_constraint.c]: let us create a function in query env and use it here
if (useTempOidBuffer(constraintNamespace))

all condition can go inside useTempOidBuffer

Comment (time-change) [reply src/backend/catalog/pg_constraint.c]: done

**Review by lejaokri — APPROVED** [review]

File: src/backend/catalog/heap.c [hunk 5]
```diff
@@ -72,6 +72,7 @@
 #include "storage/predicate.h"
 #include "utils/builtins.h"
 #include "utils/fmgroids.h"
+#include "utils/guc.h"
```
Comment (surendravishnoi) [inline src/backend/catalog/heap.c:75]: do we need this?

File: src/backend/catalog/pg_constraint.c [hunk 6]
```diff
@@ -22,16 +22,20 @@
 #include "catalog/catalog.h"
 #include "catalog/dependency.h"
 #include "catalog/indexing.h"
+#include "catalog/namespace.h"
```
Comment (surendravishnoi) [inline src/backend/catalog/pg_constraint.c:25]: do we need all these headers?

**Review by surendravishnoi — APPROVED** [review]

**Review by 2jungkook — APPROVED** [review]
